cmd/kubeadm/app/util/version.go

```diff
@@ -96,10 +96,6 @@ func kubernetesReleaseVersion(version string, fetcher func(string, time.Duration
 		url := fmt.Sprintf("%s/%s.txt", bucketURL, versionLabel)
 		body, err := fetcher(url, getReleaseVersionTimeout)
 		if err != nil {
-			// If the network operaton was successful but the server did not reply with StatusOK
-			if body != "" {
-				return "", err
-			}
```

Comment:
for the attached issue:
> The CA generation failed because of 502 bad gateway returned and the response body isn't empty when it's trying to fetch stable version from internet. 

what does the body contain?

should we just handle StatusOK?
@kad had comments that we should not do that, thus we ended with the `body` check.

Reply:
it was specifically 404 which should be leading to propagation of error, but code was written in a bit simpler manner: "if not StatusOK"... Probably we should do a bit better handling for 5xx statuses and allow fallback in that case, similarly as with detection of "air gap" scenarios. I  need to dig a bit more in the code to see how to do it better.

Reply:
> what does the body contain?
Since the body is not logged anywhere, I don't have the body at the moment. Unless we can reproduce the issue

> should we just handle StatusOK?
Looks the StatusOk is already handled [here](https://github.com/kubernetes/kubernetes/blob/v1.13.7/cmd/kubeadm/app/util/version.go#L176)? So that `fetchFromURL` returns error for any non Ok status

Reply:
@kad @leakingtapan 
we should probably just handle 5** as a quick fix?

long term we discussed that we should make the "fetch remote version code" on demand only (if a label is passed) and use the local version by default.

Reply:
Here are my impressions ATM:
1. 4** and 5** codes should always be considered errors (especially 5**).
2. The body of everything else, but 200, is unreliable. Even if it contains a version, it may not be a Kubernetes version.
3. We need a test case that handles this.

Reply:
@rosti 
Do you mean to add a test case for `kubernetesReleaseVersion()`, and the test case provides fetcher that returns non-200 code and non-empty body?

Reply:
> The body of everything else, but 200, is unreliable. Even if it contains a version, it may not be a Kubernetes version.

an empty body should fail here:
https://github.com/kubernetes/kubernetes/blob/3e2f9b8fbc7b27fe5874750a587d26a2154ecc4b/cmd/kubeadm/app/util/version.go#L118

> and the test case provides fetcher that returns non-200 code and non-empty body?

if we want to unit test fetchFromURL() we need to break it into two parts:
- creates a client (real / fake for unit tests https://golang.org/pkg/net/http/httptest/).
- separate function X that calls Get() on the client.

the function X can be unit tested:
- empty body and no-error
- for body that does not parse as version and no-error
- no-error, body that has version
...

Reply:
We should have a test, where something valid and typical (such as `stable-1`) is supplied, but the fetcher fails with an error. In that case package version should be used and no error returned.
Unfortunately, that being an unit test, there is no package version (just some junk).

Reply:
@rosti 4xx is an error for sure - we asked something that is definitely not valid on the server side. 5xx are sever side errors, not necessarily based on our request, so potentially ok to fallback.

Reply:
> @rosti 4xx is an error for sure - we asked something that is definitely not valid on the server side. 5xx are sever side errors, not necessarily based on our request, so potentially ok to fallback.

If im understanding correctly, we are going to fallback whenever there is an error from `fetcher()`. Am I correct? I'm a bit confused on how does the body matters in the case of error

Reply:
@kad I am not very convinced, that a valid version can be obtained from a 5** response body, but at least, we can try to parse that upon error and fall back to the error if no valid version is found.
So how about settling on that?

Reply:
I'm confused now.

Let me know how I can help shape it up. :)

Reply:
@rosti @neolit123 @kad 

I suggest not to be too complicated.  As mentioned by @leakingtapan , we going to fallback when there is an error from `fetcher()`.  We don't need to consider the error code is 4xx or 5xx, and if there is an error that makes no sense to parse the body.

I don't think it's worth to break `fetchFromURL()` into two parts, just for the low-value test case. 

I look forward to your final conclusion.

Reply:
@kad please clarify your points so we can move the fix forward

Reply:
we discussed this problem again today with @rosti 
@RainbowMango please fix the failing unit tests https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/80024/pull-kubernetes-bazel-test/1151838526606675968

and we can merge this PR, follow up PRs can still be accepted (cc @kad).

Reply:
@neolit123 
Done.